src/pyk/ktool/kompile.py

```diff
@@ -1,327 +1,307 @@
-    syntax_module: str | None,
-    include_dirs: Iterable[Path],
-    md_selector: str | None,
-    hook_namespaces: Iterable[str],
-    emit_json: bool,
-    gen_bison_parser: bool,
-    bison_parser_library: bool,
-    debug: bool = False,
-    post_process: str | None,
-    read_only: bool = False,
-    llvm_kompile_type: LLVMKompileType | None = None,
-    llvm_kompile_output: str | None = None,
-    opt_level: int | None,
-    ccopts: Iterable[str],
-    no_llvm_kompile: bool,
-    enable_search: bool,
-    enable_llvm_debug: bool = False,
-    concrete_rules: Iterable[str],
-) -> list[str]:
-    args = list(command) + [str(main_file)]
-
-    if output_dir:
-        args += ['--output-definition', str(output_dir)]
-
-    if backend:
-        args += ['--backend', backend.value]
-
-    if main_module:
-        args.extend(['--main-module', main_module])
-
-    if syntax_module:
-        args.extend(['--syntax-module', syntax_module])
-
-    for include_dir in include_dirs:
-        args += ['-I', str(include_dir)]
-
-    if md_selector:
-        args.extend(['--md-selector', md_selector])
-
-    if hook_namespaces:
-        args.extend(['--hook-namespaces', ' '.join(hook_namespaces)])
-
-    if emit_json:
-        args.append('--emit-json')
-
-    if gen_bison_parser:
-        args.append('--gen-bison-parser')
-
-    if bison_parser_library:
-        args.append('--bison-parser-library')
-
-    if debug:
-        args.append('--debug')
-
-    if post_process:
-        args.extend(['--post-process', shlex.quote(post_process)])
-
-    if read_only:
-        args.append('--read-only-kompiled-directory')
-
-    if llvm_kompile_type is not None:
-        args.extend(['--llvm-kompile-type', llvm_kompile_type.value])
-
-    if llvm_kompile_output is not None:
-        args.extend(['--llvm-kompile-output', llvm_kompile_output])
-
-    if opt_level:
-        args.append(f'-O{opt_level}')
+class KompileBackend(Enum):
+    LLVM = 'llvm'
+    HASKELL = 'haskell'
+    KORE = 'kore'
 
-    for ccopt in ccopts:
-        args += ['-ccopt', ccopt]
 
-    if no_llvm_kompile:
-        args.append('--no-llvm-kompile')
+class Kompile(ABC):
+    base_args: KompileArgs
+
+    @staticmethod
+    def from_dict(dct: Mapping[str, Any]) -> Kompile:
+        backend = KompileBackend(dct.get('backend') or 'llvm')
+        common_args = {key: value for key, value in dct.items() if key in COMMON_ARGS}
+        base_args = KompileArgs(**common_args)
+        match backend:
+            case KompileBackend.HASKELL:
+                haskell_args = {key: value for key, value in dct.items() if key in HASKELL_ARGS}
+                return HaskellKompile(base_args, **haskell_args)
+            case KompileBackend.LLVM:
+                llvm_args = {key: value for key, value in dct.items() if key in LLVM_ARGS}
+                return LLVMKompile(base_args, **llvm_args)
+            case _:
+                raise ValueError(f'Unsupported backend: {backend.value}')
```

Comment:
Can we check that if the backend is set to `llvm`, that no haskell args are used, for example?

```suggestion
    def from_dict(dct: Mapping[str, Any]) -> Kompile:
        backend = KompileBackend(dct.get('backend') or 'llvm')
        common_args = {key: value for key, value in dct.items() if key in COMMON_ARGS}
        base_args = KompileArgs(**common_args)
        llvm_args = {key: value for key, value in dct.items() if key in LLVM_ARGS}
        haskell_args = {key: value for key, value in dct.items() if key in HASKELL_ARGS}
        match backend:
            case KompileBackend.HASKELL:
                assert len(llvm_args) == 0
                return HaskellKompile(base_args, **haskell_args)
            case KompileBackend.LLVM:
                assert len(haskell_args) == 0
                return LLVMKompile(base_args, **llvm_args)
            case _:
                raise ValueError(f'Unsupported backend: {backend.value}')
```

For example (though may not be as good as a version that actually prints an error message with the offending args listed).

Reply:
Done.